core/Pos/Crypto/SignTag.hs

```diff
@@ -0,0 +1,56 @@
+module Pos.Crypto.SignTag
+       ( SignTag(..)
+       , signTag
+       ) where
+
+import           Universum
+
+import qualified Data.ByteString.Lazy as BSL
+import qualified Data.Text.Buildable
+import           Formatting           (bprint, shown)
+
+import qualified Pos.Binary.Class     as Bi
+import           Pos.Core.Constants   (protocolMagic)
+
+-- | To protect agains replay attacks (i.e. when an attacker intercepts a
+-- signed piece of data and later sends it again), we add a tag to all data
+-- that we sign. This ensures that even if some bytestring can be
+-- deserialized into two different types of messages (A and B), the attacker
+-- can't take message A and send it as message B.
+--
+-- We also automatically add the network tag ('protocolMagic') whenever it
+-- makes sense, to ensure that things intended for testnet won't work for
+-- mainnet.
+data SignTag
+    -- | Transaction input:    @TxSigData@
+    = SignTxIn
+    -- | Vss certificate:      @(AsBinary VssPublicKey, EpochIndex)@
+    | SignVssCert
+    -- | Update proposal:      @UpdateProposalToSign@
+    | SignUSProposal
+    -- | Commitment:           @(EpochIndex, Commitment)@
+    | SignCommitment
+    -- | Update proposal vote: @(UpId, Bool)@
+    | SignUSVote
+    -- | Main block:           @MainToSign@
+    | SignMainBlock
+    deriving (Eq, Ord, Show, Generic, Typeable)
+
+-- TODO: it would be nice if we couldn't use 'SignTag' with wrong
+-- types. Maybe something with GADTs and data families?
```

Comment:
:+1: for GADT. You can create issue and I can try accomplish this :) Or even try in this branch later.